compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -371,6 +381,66 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         );
     }
 
+    // The various `impl<T: FnPtr> Trait for T` in libcore are more like builtin impls for all function items
+    // and function pointers and less like blanket impls. Rejecting them when they can't possibly apply (because
+    // the obligation's self-type does not implement `FnPtr`) avoids reporting that the self type does not implement
+    // `FnPtr`, when we wanted to report that it doesn't implement `Trait`.
+    #[instrument(level = "trace", skip(self), ret)]
+    fn reject_fn_ptr_impls(
+        &self,
+        impl_def_id: DefId,
+        obligation: &TraitObligation<'tcx>,
+        self_ty: Ty<'tcx>,
+    ) -> bool {
+        // Let `impl<T: FnPtr> Trait for Vec<T>` go through the normal rejection path.
+        if !matches!(self_ty.kind(), ty::Param(..)) {
+            return false;
+        }
+        let Some(fn_ptr_trait) = self.tcx().lang_items().fn_ptr_trait() else {
+            return false;
+        };
+        for &(predicate, _) in self.tcx().predicates_of(impl_def_id).predicates {
+            let ty::PredicateKind::Clause(ty::Clause::Trait(pred))
+                = predicate.kind().skip_binder() else { continue };
+            if fn_ptr_trait != pred.trait_ref.def_id {
+                continue;
+            }
+            trace!(?pred);
+            // Not the bound we're looking for
+            if pred.self_ty() != self_ty {
+                continue;
+            }
+
+            match obligation.self_ty().skip_binder().kind() {
+                // Fast path to avoid evaluating an obligation that trivally holds.
+                // There may be more bounds, but these are checked by the regular path.
+                ty::FnPtr(..) | ty::FnDef(..) => continue,
```

Comment:
`FnDef` does not impl `FnPtr` :thinking: is that for diagnostics in the `fn_def: Debug` case so that we emit the error `fn_def: FnPtr` due to the single applicable `impl<T: FnPtr> Debug for T` impl?